pkg/gatewayserver/gatewayserver.go

```diff
@@ -972,6 +990,36 @@ func (gs *GatewayServer) handleLocationUpdates(ctx context.Context, conn connect
 	}
 }
 
+// handleVersionInfoUpdates updates gateway attributes with version info.
+// This function runs exactly once; only for the first status message of each connection, since version information does (should) not change within the same connection.
+func (gs *GatewayServer) handleVersionInfoUpdates(ctx context.Context, conn connectionEntry) {
+	select {
+	case <-ctx.Done():
+		return
+	case <-conn.VersionInfoChanged():
+		status, _, ok := conn.StatusStats()
+		if ok && status.Versions["model"] != "" && status.Versions["firmware"] != "" {
+			if status.Versions["model"] == conn.Gateway().Attributes["model"] && status.Versions["firmware"] == conn.Gateway().Attributes["firmware"] {
+				return
+			}
+			attributes := map[string]string{
+				"model":    status.Versions["model"],
+				"firmware": status.Versions["firmware"],
+			}
+			d := random.Jitter(gs.config.UpdateVersionInfoDelay, 1)
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(d):
+			}
+			err := gs.entityRegistry.UpdateAttributes(conn.Context(), *conn.Gateway().Ids, *&conn.Gateway().Attributes, attributes)
```

Comment:
`*&` does not copy the object. Make a clone explicitly before the call, or inside `UpdateAttributes`.